### PR TITLE
Support finding data files for sources without CONTROL keys

### PR DESCRIPTION
**Pull request by philsmt**
The REMI reconstruction in CAL has created virtual control sources with only `RUN` keys (and thus no `CONTROL` keys) for a while now. This has mostly worked with EXtra-data so far, but while trying to reduce data I noticed the metadata properties like `SourceData.aggregator` failing in `_get_first_source_file`. This happens because `SourceData.one_key()` (rightfully so!) returns `None` in this case, which the downstream parts cannot handle.

This PR tries to add support for this situation. I keep it a draft for now in case there are objections, and add tests if there are none. The pieces are:

- 8f06485628c1dfe4e3a4325fca78e6df971f947a: In hindsight now completely orthogonal to this PR, but while understanding the code it seemed sensible to move this implementation to `DataChunk`, as it solely operates on these objects. I'm happy to remove this commit if there are objections.
- 54fc91567fda992809e5f39342b3f4f186679807: Raise a `TypeError` in `SourceData.__getitem__` for unexpected dates (here `None`) analogous to `DataCollection.__getitem__`.
- 7e7d30b862e192a624d88ef423c430ddc37c22c9: Add support for `CONTROL`-less control sources in `_get_first_source_file` based on the exception added in the prior commit.

### CODE REVIEW

File: extra_data/sourcedata.py [hunk 1]
```diff
@@ -102,11 +105,28 @@ def _ipython_key_completions_(self):
...
             sample_path = first_kd.source_file_paths[0]
+
+        except TypeError:
+            for file in self.files:
+                group = file.file['RUN'].get(self.source, None, getlink=True)
+                if group is not None:
+                    break
+            else:
+                # Should this be possible?
+                raise FileStructureError(f'No CONTROL or RUN keys present '
```
Comment (takluyver) [inline extra_data/sourcedata.py:122]: This error message could be confusing if it occurs on an instrument source. It shouldn't be possible in that case, but maybe we should check `self.is_control` before going into the fallback looking for RUN.

File: extra_data/sourcedata.py [hunk 2]
```diff
@@ -5,7 +5,8 @@
 import h5py
 import numpy as np
 
-from .exceptions import MultiRunError, NoDataError, PropertyNameError
+from .exceptions import (MultiRunError, NoDataError, PropertyNameError,
+                         FileStructureError)
```
[the remark this answers is not in the document]
Comment (philsmt) [reply extra_data/sourcedata.py:9]: Uhm, thanks for nitpicking without being asked!

File: extra_data/sourcedata.py [hunk 3]
```diff
@@ -102,11 +105,28 @@ def _ipython_key_completions_(self):
...
-            # This property is an empty list if no trains are selected.
+            # May throw TypeError if there are only RUN keys.
+            first_kd = self[self.one_key()]
+
+            # May throw IndexError if no trains are selected.
             sample_path = first_kd.source_file_paths[0]
+
+        except TypeError:
+            for file in self.files:
+                group = file.file['RUN'].get(self.source, None, getlink=True)
```
[the remark this answers is not in the document]
Comment (philsmt) [reply extra_data/sourcedata.py:117]: TIL it's a `HardLink` for actual groups!

**Comment by philsmt** [comment]
> I imagine this will be somewhat tedious, but can we have a test for the fallback? It probably means adding a new device to the mockdata machinery.

Definitely, just wanted to start discussion before.

Actually we already have a [mock version](https://github.com/European-XFEL/EXtra/blob/master/tests/mockdata/dld.py) oif the device causing this, but now it'll finally require us to fix https://github.com/European-XFEL/EXtra-data/issues/429 before.

**Comment by philsmt** [comment]
be941b5e770093dc854d3ce22d490bf1ec4e6d14 now fixes https://github.com/European-XFEL/EXtra-data/issues/429 while keeping compatibility with the old API. In principle we can also change all sites where it is used if you prefer (but this must include EXtra!)

883a94f38c32e061fb78910879e558dce37d970d is another fix to support key-less sources in `FileAccess.get_one_key`, which lurked in a cache branch and was triggered by the test.

Otherwise I added all suggestions above.

**Comment by takluyver** [comment]
I think we can go ahead with this when you're ready.

I also ran into a related failure in `.plot_missing_data()`: this called `SourceData.data_counts()`, which does `self[self.one_key(index_group)]`. `one_key()` returns None, and `srcdata[None]` fails unexpectedly. Is it practical to catch this case in this PR as well?

**Comment by philsmt** [comment]
Well spotted, I added a fix in b9d93a482ddb8d1eef8b30c11ea3f34cb765f678

**Comment by takluyver** [comment]
Thanks! I guess we might want to skip such sources in `.plot_missing_data()` as well, since otherwise they'll always show up as missing all trains.

**Comment by philsmt** [comment]
I was considering that, though somehow I would expect `SourceData.data_counts()` to return an all-zero array given that, well, there isn't data for any train. That in turn is then also true for `.plot_missing_data()`. This is somewhat reminiscent of the missing data intentionally vs accidentally in the original `.plot_missing_data()` PR.

But note I don't feel strongly and am fine with indeed removing it.

**Comment by takluyver** [comment]
I think it makes sense for `.data_counts()` to say there are 0 entries per train, but I don't think it's helpful for those sources to show up in `.plot_missing_data()`, because we expect them to have no control data.

Although thinking more about it, if you do a `.select(..., require_all=True)` including such a source, presumably it will deselect all trains, which isn't really useful either. Maybe `.data_counts()` should claim that there's 1 entry per train, like any other CONTROL source, it just happens to have no keys you can get data for.

**Comment by philsmt** [comment]
Valid points, let's get some other view points in here? @tmichela @JamesWrigley

**Comment by philsmt** [comment]
Discussing with Thomas offline, `counts = SourceData.data_counts()` *should* return zero, as its sum is understood as the number of rows any key result would have.
Would it be a compromise to simply omit sources with no keys in `DataCollection.plot_missing_data`?

**Comment by philsmt** [comment]
Thanks for review!

**Comment by philsmt** [comment]
@takluyver Out of curiosity, do you understand the *project* coverage test failing and the *patch* not? It's the entirety of EXtra-data below our coverage threshold?
